syntax_checkers/rust/rustc.vim

```diff
@@ -10,11 +10,29 @@ if exists("g:loaded_syntastic_rust_rustc_checker")
 endif
 let g:loaded_syntastic_rust_rustc_checker = 1
 
+if !exists('g:rustc_syntax_only')
+    let g:rustc_syntax_only = 1 "Keep the fast behaviour by default
+endif
+
 let s:save_cpo = &cpo
 set cpo&vim
 
 function! SyntaxCheckers_rust_rustc_GetLocList() dict
-    let makeprg = self.makeprgBuild({ 'args': '-Zparse-only' })
+    let compiler_params = g:rustc_syntax_only ? '-Zparse-only' : '-Zno-trans'
+    let cwd = '.' " Don't change cwd as default
+    let cargo_toml_path = findfile('Cargo.toml', '.;')
+    if empty(cargo_toml_path) " Plain rs file, not a crate
+        let makeprg = self.makeprgBuild({
+            \ 'exe': 'rustc',
+            \ 'args': compiler_params})
+    else " We are inside a crate
+        let makeprg = self.makeprgBuild({
+            \ 'exe': 'cargo',
+            \ 'args': 'rustc ' . compiler_params,
+            \ 'fname': '' })
+        " Change cwd to the root of the crate
+        let cwd = fnamemodify( cargo_toml_path, ':p:h')
```

Comment:
this line currently doesn't work inside a workspace, for it needs to find workspace's Cargo.toml (which is ../ relative to what it currently finds), so reported filenames in the errors are like:
`  --> recompile_self/build.rs:32:3`
but cwd is `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage/recompile_self`
when it should be
`/home/xftroxgpx/build/2nonpkgs/rust.stuff/rustlearnage`
for vim/syntastic to actually show the errors upon save.

I don't know of a way to detect if we're inside a workspace now...